### ISSUE TITLE
Performance improvement

### ISSUE THREAD

**Issue by podtserkovskiy** [SpectoLabs/hoverfly#788]
Hi! The hoverfly does decoding response from a slice of bytes to a struct on every request. It causes high consumption of CPU and huge number allocations of memory. Can we avoid this effects by additional cache layer or by another way? Performance of hoverfly is valuable for us because we want to try use hoverfly for performance testing of microservices as a mock for external services.

[pprof svg files](https://github.com/SpectoLabs/hoverfly/files/2609044/Archive.zip)

**Comment by tommysitu** [comment]
Hi @podtserkovskiy, before I dig into the code, can you let me know what your hoverfly configuration was when you were doing the profiling, please?

**Comment by podtserkovskiy** [comment]
I ran hoverfly as webserver with a simple config like the code below, maybe response body was greater about 10KB.
```
{
  "data": {
    "pairs": [
      {
        "response": {
          "status": 200,
          "body": "5iNe8dxWH5Ca8pZqAfEHv3rgC0SsvKNLu6o3K",
          "encodedBody": false,
          "headers": {
            "Accept-Ranges": [
              "bytes"
            ],
            "Cache-Control": [
              "max-age=3600"
            ],
            "Connection": [
              "keep-alive"
            ],
            "Content-Type": [
              "text/html; charset=utf-8"
            ],
            "Date": [
              "Tue, 30 May 2017 13:23:09 GMT"
            ],
            "Etag": [
              "\"82b6bafbc0c4af5e3886d07802f4b62d\""
            ],
            "Hoverfly": [
              "Was-Here"
            ],
            "Last-Modified": [
              "Fri, 19 May 2017 10:59:12 GMT"
            ],
            "Server": [
              "nginx"
            ],
            "Strict-Transport-Security": [
              "max-age=31556926"
            ],
            "Transfer-Encoding": [
              "chunked"
            ],
            "Vary": [
              "Accept-Encoding"
            ]
          }
        },
        "request": {
          "path": [
            {
              "matcher": "exact",
              "value": "/bar"
            }
          ],
          "method": [
            {
              "matcher": "exact",
              "value": "GET"
            }
          ],
          "query": {},
          "body": [
            {
              "matcher": "exact",
              "value": ""
            }
          ]
        }
      }
    ],
    "globalActions": {
               "delays": [
            {
               "urlPattern": ".",
               "httpMethod": "",
               "delay": 500
            }
         ]
    }
  },
  "meta": {
    "schemaVersion": "v5",
    "hoverflyVersion": "v0.17.0",
    "timeExported": "2017-05-30T14:23:44+01:00"
  }
}
```

**Comment by tommysitu** [comment]
What about using https://github.com/patrickmn/go-cache to cache the struct directly, without encoding/decoding. What do you think @podtserkovskiy?

**Comment by podtserkovskiy** [comment]
I've never used go-cache package, but I think it's a good idea to use cache without encoding/decoding.

**Comment by tommysitu** [comment]
Thanks @podtserkovskiy I will give it a try.

**Comment by tommysitu** [comment]
Using go-cache seems to solve the problem, my benchmark shows `150574 ns/op` before the change, and `5729 ns/op` after. It's a significant improvement.

**Comment by tommysitu** [comment]
It turns out that using LRU cache is better: https://github.com/hashicorp/golang-lru, it will cap the memory usage during load test.

**Comment by podtserkovskiy** [comment]
Thank you!

**Comment by tommysitu** [comment]
This is fixed in v1.0.0-rc.2.
